File: acwing/算法提高课/第一章动态规划/1018.最低通行费/1018.最低通行费.cpp
#include <iostream>
#include <algorithm>
#include <cstring>

using namespace std;

const int N = 110, INF = 1e9;

int f[N][N];
int w[N][N];

int main()
{
    int n;
    scanf("%d", &n);
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            scanf("%d", &w[i][j]);
        }
    }
    // memset(f, INF, sizeof(f));
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            if (i == 1 && j == 1)
            {
                f[i][j] = w[i][j];
            }
            else
            {
                f[i][j] = INF;
                if (i > 1)
                    f[i][j] = min(f[i][j], f[i - 1][j] + w[i][j]);
                if (j > 1)
                    f[i][j] = min(f[i][j], f[i][j - 1] + w[i][j]);
            }
        }
    }
    printf("%d\n", f[n][n]);
}

File: acwing/算法提高课/第一章动态规划/1018.最低通行费/1018.最低通行费.py
N = 110

f = [[float("inf") for _ in range(N)] for _ in range(N)]

w = []

n = int(input())

for _ in range(n):
    w.append(list(map(int, input().split())))


for i in range(1, n + 1):
    for j in range(1, n + 1):
        if i == 1 and j == 1:
            f[i][j] = w[i - 1][j - 1]
        if i > 1:
            f[i][j] = min(f[i][j], f[i - 1][j] + w[i - 1][j - 1])
        if j > 1:
            f[i][j] = min(f[i][j], f[i][j - 1] + w[i - 1][j - 1])

print(f[n][n])
